localstack/services/s3/provider.py

```diff
@@ -802,7 +804,28 @@ def upload_part(self, context: RequestContext, request: UploadPartRequest) -> Up
                 ArgumentValue=part_number,
             )
 
-        part = body.read() if (body := request.get("Body")) else b""
+        body = request.get("Body")
```

Comment:
It feels like this decoding could actually be done automatically, similar to the GZIP decoding, right in the handler chain in a wrapping generator.
For the GZIP handling we explicitly exclude S3 because it actually needs the raw data because it directly stores it and (against the HTTP spec) the checksums are calculated for the gzip encoded content instead of the raw content.

However, here it seems that S3 always decodes the body for all operations, where it would make sense to have an explicit decoder in the handler chain. What do you think?

Reply:
I see what you mean, in the moto provider it makes sense because we read the whole payload in memory to then set it directly. However, in the new provider, it means we would iterate over the stream twice, once to decode it, then once to set it. Right now, in the stream provider, we're both decoding and validating checksum/etag at the same time, so we iterate over it only once.

I see the value of having the handler chain doing it, especially with the current implementation (it could parse the trailing headers and set it in the requests, etc), but it would also mean more work I think?

Reply:
The separation of concern between the decoding and the checksum validation would be great as well. I really like this idea, it's too bad moto won't let use implement it, I might be able to patch it though, and remove their decoding.

Reply:
If it would be problematic with the current moto integration, it is perfectly fine to move this to the content decoder in the handler chain in another iteration :)

Reply:
Perfect, but thanks a lot for the idea, it makes a lot of sense to separate it. Just need to find the right implementation now 😄